### PR TITLE
[SIL Diagnostics] Warn when assigning an integer literal to a variable of floating-point type results in loss of precision.

### PR DESCRIPTION
**Pull request by ravikandhadai**
This feature was requested in the radar issue: <rdar://15224041>.

### CODE REVIEW

**Comment by ravikandhadai** [comment]
The warnings are disabled when the conversion happens through explicit constructor call. E.g. 

`$ let f3: Float = 16777217` 
`warning: '16777217' loses precision during conversion to 'Float'`
`$ Float(16777217)` // no warning

**Comment by ravikandhadai** [comment]
@swift-ci please smoke test

**Comment by stephentyrone** [comment]
Would it be more clear if we tell the user what value actually results? Maybe something along the lines of: "Warning: '16777217' is not exactly representable as 'Float'; 'f3' has value '16777216'.

**Comment by ravikandhadai** [comment]
@swift-ci please smoke test

**Comment by ravikandhadai** [comment]
@stephentyrone  Sounds good to me. I have pushed  a commit that prints the new value in the warning messages. E.g.

$ let f3: Float = 16777217
warning: '16777217' is not exactly representable as 'Float'; it becomes '16777216'

**Comment by stephentyrone** [comment]
Nice! If you update the PR with conflicts resolved, I can review later today or tomorrow.

File: test/SILOptimizer/diagnostic_constant_propagation.swift [hunk 1]
```diff
@@ -185,49 +185,53 @@ func testConvertOverflow() {
...
-  // IEEE binary32 max value = 2^128 * (2^23-1)/2^23
-  var _ /*float32_max*/                     : Float32 = (340282326356119256160033759537265639424)
-  var _ /*float32_max_p1*/                  : Float32 = (340282326356119256160033759537265639425)
 
-  // 2^128 * (2^25-1)/2^25 - 1
-  var _ /*float32_max_not_yet_overflow*/    : Float32 = (340282356779733661637539395458142568447)
-  // 2^128 * (2^25-1)/2^25
+  var _ /*float32_max*/                     : Float32 = (340282346638528859811704183484516925440) // in hexfloat = 0x1.fffffep127
+
+  var _ /*float32_near_max_but_imprecise*/  : Float32 = (340282326356119256160033759537265639425) // expected-warning {{'340282326356119256160033759537265639425' is not exactly representable as 'Float32' (aka 'Float'); it becomes '3.40282326E+38'}}
```
Comment (jrose-apple) [inline test/SILOptimizer/diagnostic_constant_propagation.swift]: We probably want to be printing these with full precision, to make it clear what the difference is.

Comment (ravikandhadai) [reply test/SILOptimizer/diagnostic_constant_propagation.swift]: @jrose-apple Yeah sure, let me make that change.

**Comment by ravikandhadai** [comment]
@jrose-apple Done. Now the new float value is displayed like an integer to make the imprecision obvious E.g. 
`var _ : Float32 = (340282326356119256160033759537265639425)`
`> warning: '340282326356119256160033759537265639425' is not exactly representable as 'Float32' (aka 'Float'); it becomes '340282326356119256160033759537265639424'`

@stephentyrone  I have fixed the merged conflicts. Thanks for your help with this.

**Comment by ravikandhadai** [comment]
@swift-ci please smoke test

**Review by stephentyrone — APPROVED** [review]
LGTM, thanks!

**Comment by ravikandhadai** [comment]
@swift-ci please test and merge

**Comment by ravikandhadai** [comment]
@swift-ci please test and merge

**Comment by shahmishal** [comment]
@swift-ci please test and merge

**Comment by ravikandhadai** [comment]
@swift-ci please test and merge

**Comment by ravikandhadai** [comment]
@swift-ci please test and merge

**Pull request merged**